File: incre-tests/autolifter/single-pass/largest_peak.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = 0;
Config SampleIntMax = 3;

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

largest_peak = fix (
  \f: Int -> List -> Int. \cmo: Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} ->
    let cmo = if > h 0 then + cmo h else 0 in
    max cmo (f cmo t)
  end
) 0;

/*pos_prefix = fix (
  \f: List -> List. \xs: List.
  match xs with
    nil _ -> nil unit
  | cons {h, t} -> if > h 0 then cons {h, f t} else nil unit
  end
);*/

main = single_pass largest_peak;

File: ml-benchmark/autolifter/single-pass/largest_peak.ml
type lst =
  | Cons of int * lst
  | Nil

let single_pass v =
  let rec run xs =
    match xs with
    | Nil -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let inf = 100

let max a b = if a < b then b else a

let rec largest_peak_f cmo xs =
  match xs with
  | Nil -> 0
  | Cons (h, t) ->
    let cmo = if h > 0 then cmo + h else 0 in
    max cmo (largest_peak_f cmo t)

let largest_peak = largest_peak_f 0

let program = single_pass largest_peak